poorwsgi/digest.py

```diff
@@ -0,0 +1,351 @@
+                raise HTTPException(state.HTTP_UNAUTHORIZED, realm=realm)
+
+            if not check_token(req.authorization.get('nonce'),
+                               req.secret_key, req.user_agent,
+                               timeout=req.app.auth_timeout):
+                log.info("Digest: nonce value not match")
+                raise HTTPException(state.HTTP_UNAUTHORIZED, realm=realm,
+                                    stale=True)
+
+            if not check_credentials(req, realm, username):
+                raise HTTPException(state.HTTP_UNAUTHORIZED, realm=realm)
+
+            req.user = req.authorization['username']
+            return fun(req)
+        return handler
+    return wrapper
+
+
+def hexdigest(username, realm, password, algorithm=md5):
+    """Return digest hash value for user password.
+
+    Return algorithm(username:realm:password).hexdigest()
+    """
+    return algorithm(
+        ('%s:%s:%s' % (username, realm, password)).encode()
+        ).hexdigest()
+
+
+class PasswordMap(defaultdict):
+    """Simple memory object to store user password."""
+    def __init__(self, pathname=None):
+        super().__init__(dict)
+        self.pathname = pathname
+
+    def set(self, realm, username, digest):
+        """Add username to realm."""
+        self[realm][username] = digest
+
+    def delete(self, realm, username):
+        """Delete username from realm."""
+        return bool(self[realm].pop(username, None))
+
+    def find(self, realm, username):
+        """Return digest for username in realm if exist."""
+        if realm in self and username in self[realm]:
+            return self[realm][username]
+        return None
+
+    def verify(self, realm, username, digest):
+        """Check digest in password map."""
+        digest_ = self.find(realm, username)
+        return bool(digest_) and digest_ == digest
+
+    def load(self):
+        """Load map from file."""
+        if self.pathname is None:
+            raise RuntimeError("No pathname was set.")
+
+        with open(self.pathname) as pwfile:
+            for line in pwfile:
+                username, realm, digest = line.strip().split(':')
+                self.set(realm, username, digest)
+
+    def write(self):
+        """Write memory map dump."""
+        if self.pathname is None:
+            raise RuntimeError("No pathname was set.")
+
+        with open(self.pathname, 'w+') as pwfile:
+            for realm, vals in self.items():
+                for username, digest in vals.items():
+                    pwfile.write('%s:%s:%s\n' % (username, realm, digest))
+
+
+def get_re_type():
+    """Get password from stdin with re-type ."""
+    password = getpass('New password: ')
+    re_type = getpass('Re-type new password: ')
+    if password != re_type:
+        print('They don\'t match, sorry.')
+        return None
+    return password
+
+
+def main():
+    """Main fce for manipulation with passwordfile."""
```

Comment:
fce -> function. (Personally I know only the `fct` abbreviation).

Reply:
fce is from czech language :D